posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd

```diff
@@ -0,0 +1,339 @@
+We will also explain how to graphically demonstrate the time taken by each operation. We will use the atime package to compare and visualize the asymptotic performance (time and memory usage) of the different functions mentioned above.
+By comparing the asymptotic performance of these packages in these programming languages, We aim to provide insights into their usage and help data scientists make informed choices when it comes to data manipulation and analysis.
+
+In the ```atime::atime()``` we define the following arguments:
+The first argument, N, is a sequence of data sizes. This explores computational efficiency, which will compute time/memory usage for different data sizes (N). 
+The second argument setup is an expression that will be evaluated for each value in N, to create data of a given size.
+seconds.limit may optionally be specified. If an expression is slower than this limit for any data size, then no larger data sizes will be measured.
+Lastly, expr.list, which is the list of expressions for which time/memory usage will be measured.
+
+## Libraries
+
+```{r setup,warning=FALSE,message=FALSE}
+library(data.table)
+library(reshape2)
+library(atime)
+library(ggplot2)
+library(reticulate)
+use_python("C:/Users/amoak/AppData/Local/Programs/Python/Python312/python.exe") #If you want to reproduce, please change to the path of python on your computer.
+virtualenv_create("fm-proj")
+use_virtualenv("fm-proj", required = F)
+```
+
+```{python}
+file_path = 'data.csv'
+```
+
+# Example 1:  Writing a CSV File with data.table::fwrite() and pandas::to_csv()
+
+## fwrite: fast CSV writer
+
+Data.table provides the fwrite() function for writing data to a file while Pandas offers the to_csv() function for writing data to a CSV file.
+
+## Comparison code
+
+```{r,warning=FALSE,message=FALSE}
+write.colors <- c(
+  "data.table::fwrite" = "#D6604D",
+  "pandas.to_csv " = "#BF812D"
+)
+file_path = 'data.csv'
+n.rows <- 100
+seconds.limit <- 10
+atime.write.vary.cols <- atime::atime(
+  N = as.integer(10^seq(2, 10, by = 0.2)),
+  setup = {
+    set.seed(1)
+    input.vec <- rnorm(n.rows * N)
+    input.mat <- matrix(input.vec, n.rows, N)
+    input.df <- data.frame(input.mat) 
+    pd <- import("pandas")
+    input_df_pd <- r_to_py(input.df)
+  },
+  seconds.limit = seconds.limit,
+  "data.table::fwrite" = {
+    data.table::fwrite(input.df, tempfile(), showProgress = FALSE)
+  },
+  "pandas.to_csv" = {
+    input_df_pd$to_csv(file_path, index = FALSE)
+  }
+)
+```
+
+```{r,warning=FALSE,message=FALSE}
+refs.write.vary.cols <- atime::references_best(atime.write.vary.cols)
+plot(refs.write.vary.cols)
+```
+
+atime::references_best(): This function returns a list, which can then be visualized using plot().
```

Comment:
Why are you doing plot of references_best output? That is for comparing asymptotic complexity with references (linear, constant, quadratic) etc. In this case I guess pandas/data.table are both linear, right? In that case I don't think this plot is very interesting/informative, can you show us what it looks like?
probably the predict plot is more informative? (because it shows pandas/data.table on same axes, easier to see differences/ which one is more efficient)

Reply:
I will just take it out for now.
This is how the plot looks like for reading csv: 
![refs read vary cols](https://github.com/rdatatable-community/The-Raft/assets/143125519/c2a84c3a-b701-46a8-bdae-26b0dad3cc5f)

Reply:
actually that is interesting, since pandas looks super-linear

Reply:
addressed

Reply:
this is for writing csv:
![refs write vary cols](https://github.com/rdatatable-community/The-Raft/assets/143125519/5bf08adc-328f-4ccd-a786-db67f40de073)